### PR TITLE
fix(levm): prevent unsafe arithmetic

### LINKED ISSUE #1075: LEVM: make operations safe
We are making operations unsafely throughout the codebase, in some cases there are potential overflows. We should go through every operation made and perform them safely. We can do this by using `saturating_op`, `wrapping_op`, etc.

**Extra**: add the following lints to the `crates/vm/levm/Cargo.toml`

```toml
[lints.clippy]
arithmetic_side_effects = "deny"
panicking_overflow_checks = "warn"
manual_saturating_arithmetic = "warn"
```

### PR DESCRIPTION
**Pull request by maximopalopoli**
**Motivation**

The idea is to prevent arithmetic overflows and underflows in the levm code.

**Description**

Design desitions:
- For the gas increment I used `checked_add` and added a `ConsumedGasOverflow` error.
- For depth used `checked_add` and the `StackOverflow` error
- In the `VM::execute` method I decided to do not raise an error, because the previous behavior was to do not raise errors, so used the saturating functions
- For opcode operations used `InvalidOpcode` error

Closes #1075

### CODE REVIEW

File: crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs [hunk 1]
```diff
@@ -262,7 +282,11 @@ impl VM {
             .try_into()
             .unwrap_or(usize::MAX);
 
-        let words_copied = (size + WORD_SIZE - 1) / WORD_SIZE;
+        let words_copied = (size
+            .checked_add(WORD_SIZE)
+            .ok_or(VMError::VeryLargeNumber)?
+            .saturating_sub(1))
+            / WORD_SIZE;
```
Comment (ilitteri) [inline crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs:289]: Shouldn't we use `checked_div` here?

File: crates/vm/levm/src/opcode_handlers/system.rs [hunk 2]
```diff
@@ -43,11 +43,20 @@ impl VM {
...
 
-        let memory_byte_size = (args_offset + args_size).max(ret_offset + ret_size);
+        let memory_byte_size = (args_offset
+            .checked_add(args_size)
+            .ok_or(VMError::VeryLargeNumber)?)
+        .max(
+            ret_offset
+                .checked_add(ret_size)
+                .ok_or(VMError::VeryLargeNumber)?,
+        );
```
Comment (ilitteri) [inline crates/vm/levm/src/opcode_handlers/system.rs:53]: Nit. The parenthesis is not needed (do not commit this, it needs to be formatted).
```suggestion
        let memory_byte_size = args_offset
            .checked_add(args_size)
            .ok_or(VMError::VeryLargeNumber)?
        .max(
            ret_offset
                .checked_add(ret_size)
                .ok_or(VMError::VeryLargeNumber)?,
        );
```

File: crates/vm/levm/src/opcode_handlers/system.rs [hunk 3]
```diff
@@ -43,11 +43,20 @@ impl VM {
...
         let memory_expansion_cost = current_call_frame.memory.expansion_cost(memory_byte_size)?;
 
         let positive_value_cost = if !value.is_zero() {
-            call_opcode::NON_ZERO_VALUE_COST + call_opcode::BASIC_FALLBACK_FUNCTION_STIPEND
+            call_opcode::NON_ZERO_VALUE_COST
+                .checked_add(call_opcode::BASIC_FALLBACK_FUNCTION_STIPEND)
+                .ok_or(VMError::OverflowInArithmeticOp)?
```
Comment (ilitteri) [inline crates/vm/levm/src/opcode_handlers/system.rs:59]: What are the criteria for returning a `VMError::OverflowInArithmeticOp` and `VMError::InternalError`?

File: crates/vm/levm/src/vm.rs [hunk 4]
```diff
@@ -571,7 +638,16 @@ impl VM {
...
+        // Add gas used by the sub-context to the current one after it's execution.
+        current_call_frame.gas_used = match current_call_frame
+            .gas_used
+            .checked_add(tx_report.gas_used.into())
+        {
+            Some(gas) => gas,
+            None => {
+                return Err(VMError::ConsumedGasOverflow);
+            }
+        };
```
Comment (ilitteri) [inline crates/vm/levm/src/vm.rs:650]: Nit. We could use `.ok_or` here (it is manually implemented).

File: crates/vm/levm/src/vm.rs [hunk 5]
```diff
@@ -737,11 +813,23 @@ impl VM {
         current_call_frame: &mut CallFrame,
         gas: U256,
     ) -> Result<(), VMError> {
-        if current_call_frame.gas_used + gas > current_call_frame.gas_limit {
+        let potential_consumed_gas = match current_call_frame.gas_used.checked_add(gas) {
```
Comment (ilitteri) [inline crates/vm/levm/src/vm.rs:816]: Ditto.

File: crates/vm/levm/src/vm.rs [hunk 6]
```diff
@@ -737,11 +813,23 @@ impl VM {
...
+            }
+        };
+        if potential_consumed_gas > current_call_frame.gas_limit {
             return Err(VMError::OutOfGas);
         }
-        current_call_frame.gas_used += gas;
-        self.env.consumed_gas += gas;
+
+        current_call_frame.gas_used = potential_consumed_gas;
+        self.env.consumed_gas = match self.env.consumed_gas.checked_add(gas) {
```
Comment (ilitteri) [inline crates/vm/levm/src/vm.rs:827]: Ditto.

File: crates/vm/levm/src/opcode_handlers/system.rs (hunk 3 above)
Comment (maximopalopoli) [reply crates/vm/levm/src/opcode_handlers/system.rs:59]: The Internal's are those that should not happen in any case of execution, and the OverflowInArithmeticOp's are those that should not happen normally.

File: crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs (hunk 1 above)
Comment (maximopalopoli) [reply crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs:298]: `checked_div` only checks if denominator is zero, and WORD_SIZE is a non-zero constant

File: crates/vm/levm/src/opcode_handlers/system.rs (hunk 3 above)
Comment (maximopalopoli) [reply crates/vm/levm/src/opcode_handlers/system.rs:59]: Now I understand, should be an Internal because should not happen in any case

**Review by ilitteri — CHANGES_REQUESTED** [review]
Some created errors make sense while others don't. Below, I left some comments.

The following can be generalized as one error `ArithmeticOperationOverflow`:
- `OffsetOverflow`
- `DataSizeOverflow`
- All the current `VMError::Internal`.
- `OverflowInArithmeticOp`.

Any `GasCostOverflow`, `GasUsedOverflow`, `CreationCostIsTooHigh`, and `ConsumedGasOverflow` implies an `OutOfGas`, nevertheless, it doesn't mean that we should replace this error with `OutOfGas` but that we need to handle those cases in the corresponding layers to return an `OutOfGas` error.
For solving this, I'd suggest to make another PR (pointing to this one) abstracting the instructions gas cost logic. There should be a function `fn <instruction>(...) -> Result<integer, InternalError>` in the module `gas_cost` for each instruction that requires extra computation (that is to say that a constant is not enough for getting the cost). After this, call the corresponding function in the corresponding instruction handler, and if this function call results in an error, return an `OutOfGas` (but at the opcode handler level).

Most (if not all) of the created errors have more to do with internal errors (or arithmetic errors, as this PR says) and are not directly VM errors (they are VM errors but not VM level errors, they're a layer below). A suggestion to reorg the errors would be to create a general error at the VM level that groups all these internal errors:
```rust
#[derive(Debug, thiserror::Error)]
pub enum VMError {
    ...
    #[error("Internal error: {0}")]
    Internal(#[from] InternalError),
}

#[derive(Debug, thiserror::Error)]
pub enum InternalError {
    #[error("Overflowed when incrementing nonce")]
    NonceOverflowed,
    #[error("Underflowed when incrementing nonce")]
    NonceUnderflowed,
    #[error("Overflowed when incrementing program counter")]
    PCOverflowed,
    #[error("Underflowed when decrementing program counter")]
    PCUnderflowed,
    #[error("Arithmetic operation overflowed")]
    ArithmeticOperationOverflowed,
}
```

> [!NOTE]
> It is clear that we can generalize all the overflow errors, not doing this would be good to have more explicit errors (this is not the case though for general arithmetic operations).

File: crates/vm/levm/src/call_frame.rs [hunk 7]
```diff
@@ -114,18 +114,19 @@ impl CallFrame {
         }
     }
 
-    pub fn next_opcode(&mut self) -> Option<Opcode> {
-        let opcode = self.opcode_at(self.pc);
-        self.increment_pc();
+    pub fn next_opcode(&mut self) -> Result<Opcode, VMError> {
+        let opcode = self.opcode_at(self.pc).ok_or(VMError::InvalidOpcode);
+        self.increment_pc()?;
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/call_frame.rs]: Done! In [Change return structure in next_opcode](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1095/commits/ce39321217fefee9a99aae4f24ee9c2952850bc9)

File: crates/vm/levm/src/constants.rs [hunk 8]
```diff
@@ -1,5 +1,7 @@
 use ethereum_rust_core::U256;
 
+use crate::errors::VMError;
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/constants.rs]: Also done in [Change return structure in next_opcode](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1095/commits/ce39321217fefee9a99aae4f24ee9c2952850bc9)

File: crates/vm/levm/src/memory.rs [hunk 9]
```diff
@@ -31,54 +31,86 @@ impl Memory {
     }
 
     pub fn load(&mut self, offset: usize) -> Result<U256, VMError> {
-        self.resize(offset + 32);
+        self.resize(offset.checked_add(32).ok_or(VMError::OffsetOverflow)?);
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/memory.rs]: Done in [Replace 32 occurences for WORD_SIZE](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1095/commits/c9f91111d61180316fc5bae44d619829f613a326)

File: crates/vm/levm/src/memory.rs [hunk 10]
```diff
@@ -31,54 +31,86 @@ impl Memory {
...
         Ok(U256::from(value_bytes))
     }
 
     pub fn load_range(&mut self, offset: usize, size: usize) -> Result<Vec<u8>, VMError> {
-        self.resize(offset + size);
+        self.resize(offset.checked_add(size).ok_or(VMError::OffsetOverflow)?);
         self.data
-            .get(offset..offset + size)
+            .get(offset..offset.checked_add(size).ok_or(VMError::OffsetOverflow)?)
             .ok_or(VMError::MemoryLoadOutOfBounds)
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/memory.rs]: Done in [Refactor: Create InternalError, used it in VMError](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1095/commits/07a74a99ebf01bbb4b2a683fd59dede91e7c9175)

File: crates/vm/levm/src/memory.rs [hunk 11]
```diff
@@ -31,54 +31,86 @@ impl Memory {
...
             return Err(VMError::MemoryStoreOutOfBounds);
         }
-        self.resize(offset + len);
-        self.data
-            .splice(offset..offset + len, value.iter().copied());
+        self.resize(offset.checked_add(len).ok_or(VMError::OffsetOverflow)?);
+        self.data.splice(
+            offset..offset.checked_add(len).ok_or(VMError::OffsetOverflow)?,
+            value.iter().copied(),
+        );
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/memory.rs]: Also done in [Refactor: Create InternalError, used it in VMError](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1095/commits/07a74a99ebf01bbb4b2a683fd59dede91e7c9175)

File: crates/vm/levm/src/memory.rs [hunk 12]
```diff
@@ -31,54 +31,86 @@ impl Memory {
...
+        offset: usize,
+        value: &[u8],
+        size: usize,
+    ) -> Result<(), VMError> {
+        self.resize(offset.checked_add(size).ok_or(VMError::OffsetOverflow)?);
+        self.data.splice(
+            offset..offset.checked_add(size).ok_or(VMError::OffsetOverflow)?,
+            value.iter().copied(),
+        );
+        Ok(())
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/memory.rs]: Also done in [Refactor: Create InternalError, used it in VMError](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1095/commits/07a74a99ebf01bbb4b2a683fd59dede91e7c9175)

File: crates/vm/levm/src/memory.rs [hunk 13]
```diff
@@ -31,54 +31,86 @@ impl Memory {
...
+                    .checked_add(size)
+                    .ok_or(VMError::OffsetOverflow)?],
+        );
+
+        self.data[dest_offset
+            ..dest_offset
+                .checked_add(size)
+                .ok_or(VMError::OffsetOverflow)?]
+            .copy_from_slice(&temp);
+        Ok(())
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/memory.rs]: Done in [Refactor: Create InternalError, used it in VMError](https://github.com/lambdaclass/lambda_ethereum_rust/pull/1095/commits/07a74a99ebf01bbb4b2a683fd59dede91e7c9175)

File: crates/vm/levm/src/opcode_handlers/environment.rs [hunk 14]
```diff
@@ -204,7 +217,13 @@ impl VM {
         &mut self,
         current_call_frame: &mut CallFrame,
     ) -> Result<OpcodeSuccess, VMError> {
-        if self.env.consumed_gas + gas_cost::CODESIZE > self.env.gas_limit {
+        if self
+            .env
+            .consumed_gas
+            .checked_add(gas_cost::CODESIZE)
+            .ok_or(VMError::ConsumedGasOverflow)?
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/opcode_handlers/environment.rs]: Done in https://github.com/lambdaclass/lambda_ethereum_rust/pull/1095/commits/07a74a99ebf01bbb4b2a683fd59dede91e7c9175#diff-7e4ffc8f228de9a97e20b41a2e74215249ef36388414a31ca0806a8a9f700a84L223

File: crates/vm/levm/src/vm.rs [hunk 15]
```diff
@@ -436,14 +456,29 @@ impl VM {
...
 
             // GInitCodeword * number_of_words rounded up. GinitCodeWord = 2
             let number_of_words = self.call_frames[0].calldata.chunks(32).len() as u64;
-            report.gas_used += number_of_words * 2;
+            report.gas_used = report
+                .gas_used
+                .checked_add(
+                    number_of_words
+                        .checked_mul(2)
+                        .ok_or(VMError::VeryLargeNumber)?,
```
[the remark this answers is not in the document]
Comment (maximopalopoli) [reply crates/vm/levm/src/vm.rs]: Done [here](https://github.com/lambdaclass/lambda_ethereum_rust/commit/07a74a99ebf01bbb4b2a683fd59dede91e7c9175#diff-e20d30a6efb08393eee586d44f3bc4a8df6cbd53ec8ea5bb4d5def88c9487de1L495).

File: crates/vm/levm/src/memory.rs [hunk 16]
```diff
@@ -25,61 +25,54 @@ impl Memory {
...
-        );
+        let size_to_store = offset.checked_add(len).ok_or(VMError::Internal(
+            InternalError::ArithmeticOperationOverflow,
+        ))?;
+        /*
+               let data_len = self.data.len();
+               if data_len < offset || data_len < size_to_store {
+                   return Err(VMError::MemoryStoreOutOfBounds);
+               }
+        */
```
Comment (ilitteri) [inline crates/vm/levm/src/memory.rs:71]: What is this? Should it be removed?

File: crates/vm/levm/src/opcode_handlers/environment.rs [hunk 17]
```diff
@@ -121,6 +121,24 @@ impl VM {
...
+                   let mut padded_calldata = [0u8; WORD_SIZE];
+                   let data_len_to_copy = calldata.len().min(WORD_SIZE);
+
+                   padded_calldata[..data_len_to_copy].copy_from_slice(&calldata[..data_len_to_copy]);
+
+                   U256::from_big_endian(&padded_calldata)
+               } else {
+                   U256::zero()
+               };
+        */
```
Comment (ilitteri) [inline crates/vm/levm/src/opcode_handlers/environment.rs:140]: Ditto.

File: crates/vm/levm/src/memory.rs (hunk 16 above)
Comment (maximopalopoli) [reply crates/vm/levm/src/memory.rs:71]: In the merges I saw that these checks were not in the 'main' version, so I commented them, they are probably going to be eliminated.

File: crates/vm/levm/src/memory.rs [hunk 18]
```diff
@@ -114,21 +96,28 @@ impl Memory {
...
         let mut temp = vec![0u8; size];
 
+        temp.copy_from_slice(
+            self.data
+                .get(src_offset..src_copy_size)
+                .ok_or(VMError::Internal(InternalError::SlicingError))?,
+        );
+
+        // self.data[dest_offset..dest_copy_size].copy_from_slice(&temp);
+
```
Comment (ilitteri) [inline crates/vm/levm/src/memory.rs:120]:
```suggestion
```

**Review by ilitteri — APPROVED** [review]

**Pull request merged**